### PR TITLE
WIP: implement screenshot blackout

### PR DESCRIPTION
**Pull request by aslushnikov**
This is a WIP approach.

### CODE REVIEW

File: docs/src/api/class-page.md [hunk 1]
```diff
@@ -2683,6 +2683,11 @@ An object which specifies clipping of the resulting image. Should have the follo
 Hides default white background and allows capturing screenshots with transparency. Not applicable to `jpeg` images.
 Defaults to `false`.
 
+### option: Page.screenshot.blackout
+- `blackout` <[Array]<[Locator]>>
```
Comment (dgozman) [inline docs/src/api/class-page.md:2687]: Locator|FrameLocator ?

File: packages/playwright-core/src/client/page.ts [hunk 2]
```diff
@@ -457,10 +457,13 @@ export class Page extends ChannelOwner<channels.PageChannel> implements api.Page
     await this._channel.setNetworkInterceptionEnabled({ enabled: false });
   }
 
-  async screenshot(options: channels.PageScreenshotOptions & { path?: string } = {}): Promise<Buffer> {
+  async screenshot(options: channels.PageScreenshotOptions & { path?: string, blackout?: api.Locator[] } = {}): Promise<Buffer> {
```
Comment (dgozman) [inline packages/playwright-core/src/client/page.ts:460]: Use Locator|FrameLocator here

**Comment by p01** [comment]
Can we use a more inclusive term, e.g. `redact` ? The term "blackout" sounds terrible.

File: docs/src/api/class-page.md (hunk 1 above)
Comment (p01) [inline docs/src/api/class-page.md:2687]: s/blackout/hide|redact|cover|overlay|...

File: packages/playwright-core/src/server/injected/highlight.ts [hunk 3]
```diff
@@ -113,6 +113,27 @@ export class Highlight {
     this._actionPointElement.hidden = true;
   }
 
+  blackoutElements(elements: Element[]) {
+    const boxes = elements.map(e => e.getBoundingClientRect());
+    const pool = this._highlightElements;
+    this._highlightElements = [];
+    for (const box of boxes) {
+      const highlightElement = pool.length ? pool.shift()! : this._createHighlightElement();
+      highlightElement.style.backgroundColor = '#000';
```
Comment (p01) [inline packages/playwright-core/src/server/injected/highlight.ts:122]: As dark and light themes are common place, we can't hardcode a `#000` overlay color.

Instead we should add an `overlayColor` ( or whatever more inclusive name we pick for this feature ) property that defaults to a neutral or less common color e.g. `#808080` or `#F0F`

File: docs/src/api/class-page.md (hunk 1 above)
Comment (aslushnikov) [reply docs/src/api/class-page.md:2687]: We discussed this last week and so far settled upon `mask`.

File: packages/playwright-core/src/server/injected/highlight.ts (hunk 3 above)
Comment (aslushnikov) [reply packages/playwright-core/src/server/injected/highlight.ts:122]: I like the `#F0F`!

**Comment by aslushnikov** [comment]
Technical tidbits:

This PR explored the implementation via re-using the locators highlighting functionality in the protocol.
We decided against this approach and instead will explore serializing locators over the protocol to the server side.

The alternative implementation lives here: https://github.com/microsoft/playwright/pull/12072

**Comment by p01** [comment]
Thank you.

Good thinking about handling this scenario. 
```typescript
await Promise.all([
  page.locator('foo').screenshot({ mask: [ page.locator('.mask-1') ] }),
  page.locator('bar').screenshot({ mask: [ page.locator('.mask-2') ] }),
]);a
```
😦 TBH I would have no idea what to expect ( in terms of execution order ) from that code with the disabling of animations, the masks, ... if it wasn't serialized on the server side. Just never saw/wrote tests with this scenario.

**Pull request closed without merging**